src/tokenizer.rs

```diff
@@ -0,0 +1,510 @@
+}
+
+impl From<u8> for TokenType {
+    fn from(ch: u8) -> Self {
+        assert!(is_char_valid_token(ch));
+        // SAFETY: Always valid 'cause we assert it above
+        unsafe { mem::transmute(ch) }
+    }
+}
+
+impl From<TokenType> for u8 {
+    fn from(ty: TokenType) -> Self {
+        // SAFETY: Always valid 'cause it's memory repr is u8
+        unsafe { mem::transmute(ty) }
+    }
+}
+
+const fn is_char_valid_token(ch: u8) -> bool {
+    match ch {
+        _ if ch == b'&' => true,
+        _ if ch == b'*' => true,
+        _ if ch == b'@' => true,
+        _ if ch == b',' => true,
+        _ if ch == b':' => true,
+        _ if ch == b';' => true,
+        _ if ch == b'(' => true,
+        _ if ch == b')' => true,
+        _ if ch == b'[' => true,
+        _ if ch == b']' => true,
+        _ if ch == b'$' => true,
+        _ if ch == b'~' => true,
+        _ if ch == b'^' => true,
+        _ if ch == b'+' => true,
+        _ if ch == b'=' => true,
+        _ if ch == b'|' => true,
+        _ if ch == b'>' => true,
+        _ if ch == b' ' => true,
+        _ if ch == b'\'' => true,
+        _ if ch == b'"' => true,
+        _ if ch == b'/' => true,
+        _ if ch == b'!' => true,
+
+        _ if ch == b'\\' => true,
+        _ if ch == b'\r' => true,
+        _ if ch == 0x0C => true, // \f
+        _ if ch == b'\n' => true,
+        _ if ch == b'\t' => true,
+
+        // TODO(CGQAQ): Maybe not return true?
+        _ if ch == u8::MAX => true,
+        _ if ch == u8::MAX - 1 => true,
+        _ if ch == u8::MAX - 2 => true,
+
+        _ => false,
+    }
+}
+
+///
+/// Check if a char is not allowed to escape
+/// * return true if it's not allowed to escape.
+///
+#[inline(always)]
+pub(crate) const fn is_unescapable(t: TokenType) -> bool {
+    match t {
+        TokenType::Tab => true,
+        TokenType::Newline => true,
+        TokenType::Cr => true,
+        TokenType::Feed => true,
+        _ => false,
+    }
+}
+
+#[inline(always)]
+pub(crate) fn is_word_delimiter(t: u8) -> bool {
+    is_whitespace(t)
+        || t == TokenType::Ampersand.into()
+        || t == TokenType::Asterisk.into()
+        || t == TokenType::Bang.into()
+        || t == TokenType::Comma.into()
+        || t == TokenType::Colon.into()
+        || t == TokenType::Semicolon.into()
+        || t == TokenType::OpenParenthesis.into()
+        || t == TokenType::CloseParenthesis.into()
+        || t == TokenType::OpenSquare.into()
+        || t == TokenType::CloseSquare.into()
+        || t == TokenType::SingleQuote.into()
+        || t == TokenType::DoubleQuote.into()
+        || t == TokenType::Plus.into()
+        || t == TokenType::Pipe.into()
+        || t == TokenType::Tilde.into()
+        || t == TokenType::Greater.into()
+        || t == TokenType::Equals.into()
+        || t == TokenType::Dollar.into()
+        || t == TokenType::Caret.into()
+        || t == TokenType::Slash.into()
+    // others all considered not word delimiters.
```

Comment:
It would looks better use `matches`

Reply:
![image](https://user-images.githubusercontent.com/15936231/142762104-8c48e53f-fcc0-4b55-ac93-8ccfe884af8a.png)